src/core/providers/ogr/qgsogrprovider.cpp

```diff
@@ -1380,7 +1385,7 @@ QVariant QgsOgrProvider::defaultValue( int fieldId ) const
 
   QString defaultVal = mDefaultValues.value( fieldId, QString() );
   if ( defaultVal.isEmpty() )
-    return QVariant();
+    return defaultVal;
```

Comment:
I don't think this change is correct -- there's lots of existing code which checks if the returned value isValid(), and now that will return true instead of false (yet the default value **doesn't** exist)

Reply:
And wouldn't the check for `mDefaultValues.hasKey` be correct here?
Not that an empty string as default value is common, but it's at least in SQL terms it's valid.

Reply:
That was an attempt to homogenize the behavior, I will revert it and give up with the attempt.

But if you have a look to the tests, you'll see that the actual behavior of the two methods it is totally different between providers (I tested PG ogr and spatialite, but I expect that the others will not behave better).

What they return depends on a permutation of:
- the fact that it is a literal default or not
- the fact that it is a particular function/magic-value or not (CURRENT_TIMESTAMP for instance)
- which provider (and maybe which driver for OGR)
- if EvaluateDefaultValues is honored and its value

The least we can do is to document it in the API and possibly try to fix it for QGIS 4.